src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c

```diff
@@ -1210,14 +1210,18 @@ int xocl_gem_prime_mmap(struct drm_gem_object *obj, struct vm_area_struct *vma)
 	if (!obj->filp)
 		return -ENODEV;
 
+	/* Add the fake offset */
+	vma->vm_pgoff += drm_vma_node_start(&obj->vma_node);
+
 	ret = obj->filp->f_op->mmap(obj->filp, vma);
 	if (ret)
 		return ret;
+	drm_gem_object_get(obj);
 
 	fput(vma->vm_file);
-	if (!IS_ERR(xobj->dmabuf)) {
-		vma->vm_file = get_file(xobj->dmabuf->file);
-		vma->vm_ops = xobj->dmabuf_vm_ops;
+	if (!IS_ERR_OR_NULL(xobj->base.dma_buf->file) && !IS_ERR_OR_NULL(xobj->base.dev->driver->gem_vm_ops)) {
+		vma->vm_file = get_file(xobj->base.dma_buf->file);
+		vma->vm_ops = xobj->base.dev->driver->gem_vm_ops;
```

Comment:
xobj->dmabuf is used for qdma streaming buffer. It should not be removed. Please change this to
if (!IS_ERR(xobj->dmabuf)) {
         ...................
} else if (!IS_ERR_OR_NULL(xobj->base.dma_buf && !IS_ERR_OR_NULL(xobj->base.dma_buf->file)) {
vma->vm_pgoff += drm_vma_node_start(&obj->vma_node);      <<<<<<<<< move 1214 here 
 .............
}
I do not think you need to check gem_vm_ops which is hard coded in xocl_drm.c

Reply:
But xobj->dmabuf is nullptr here which caused the kernel hang in the beginning

Reply:
I agree. And I suggested to add "else if" above. So if ->dmabuf is NULL, it will check if base.dma_buf is NULL.

Reply:
line 1214 which adds fake offset should be called before f_op->mmap, so added it there.

Reply:
@houlz0507 added changes as per comments. Please review